sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -769,6 +783,13 @@ object LimitPushDown extends Rule[LogicalPlan] {
     // Push down local limit 1 if join type is LeftSemiOrAnti and join condition is empty.
     case j @ Join(_, right, LeftSemiOrAnti(_), None, _) if !right.maxRows.exists(_ <= 1) =>
       j.copy(right = maybePushLocalLimit(Literal(1, IntegerType), right))
+
+    case GlobalLimit(l, proj @ Project(projectList, child)) if canPushLimitThroughProject(proj) =>
+      Project(projectList, GlobalLimit(l, child))
+    case LocalLimit(l, proj @ Project(projectList, child)) if canPushLimitThroughProject(proj) =>
+      Project(projectList, LocalLimit(l, child))
+    case Limit(l, proj @ Project(projectList, child)) if canPushLimitThroughProject(proj) =>
+      Project(projectList, Limit(l, child))
```

Comment:
How about?
```scala
    case GlobalLimit(l, p @ Project(projectList, child))
        if child.outputSet.subsetOf(p.references) =>
      Project(projectList, GlobalLimit(l, child))
    case LocalLimit(l, p @ Project(projectList, child))
       if child.outputSet.subsetOf(p.references) =>
      Project(projectList, LocalLimit(l, child))
    case Limit(l, p @ Project(projectList, child))
        if child.outputSet.subsetOf(p.references) =>
      Project(projectList, Limit(l, child))
```

Reply:
Thanks, @wangyum. This is a really much better solution. The failed python ut should be fixed by this update.